_posts/2022-08-04-calldata-tuple-reencoding-head-overflow-bug.md

```diff
@@ -0,0 +1,219 @@
+---
+layout: post
+published: true
+title: 'Head Overflow Bug in Calldata Tuple ABI-Reencoding.'
+date: '2022-08-04'
+author: Solidity Team
+category: Security Alerts
+---
+
+On July 5, 2022, Chance Hudson ([@vimwitch](https://github.com/vimwitch)) from Ethereum Foundation discovered a bug in the Solidity code generator.
+
+The earliest affected version of the compiler is 0.5.8, which introduced ABI-reencoding of calldata structs.
+[Solidity version 0.8.16](https://github.com/ethereum/solidity/releases/tag/v0.8.16) released on August 04, 2022 provides a fix.
+
+We assigned the bug a severity of "medium".
+
+## Which Contracts are Affected?
+
+The effects of the bug manifest when a contract performs ABI-encoding of a tuple that meets **all** of the following conditions:
+1. Data to be encoded is stored in calldata.
```

Comment:
Only the statically-sized last component needs to come from calldata, right? (i.e. [test/libsolidity/semanticTests/abiEncoderV2/calldata_struct_dynamic_and_static.sol](https://github.com/ethereum/solidity/pull/13282/files#diff-46731f94379021f2c0fe9cbac446bfd7ebdd9ee8e41d993913fa70bf8ac706e7) is affected)
This point sounds to me like *all* data you want to encode has to come from calldata...

Reply:
Right. I simplified that too much. I was mostly thinking about the struct case or calling from outside of a contract and then everything sits in `calldata`. But yeah, with `abi.encode()` that's not necessarily the case.

Reply:
Also without ``abi.encode()``... this should have it as well:
```
contract C {
        function g(uint[] memory a, uint[1] memory) public returns (uint) {
                return a.length;
        }
        function f(uint[1] calldata b) public returns (uint) {
                return this.g(new uint[](3), b);
        }
}
// ----
// f(uint256[1]): 0 -> 0
```
So we can, for example, just drop current point ``1`` and just add that that statically sized array needs to be in calldata.